spotbugs/src/main/java/edu/umd/cs/findbugs/detect/DateFormatStringChecker.java

```diff
@@ -0,0 +1,183 @@
+        }
+
+        if (idx == -1) {
+            return;
+        }
+
+        String dateFormatString = (String) stack.getStackItem(idx).getConstant();
+        if (dateFormatString != null && runDateFormatRuleVerify(dateFormatString)) {
+            bugReporter.reportBug(new BugInstance(this, "FS_BAD_DATE_FORMAT_FLAG_COMBO", NORMAL_PRIORITY)
+                    .addClassAndMethod(this).addCalledMethod(this).addString(dateFormatString)
+                    .describe(StringAnnotation.FORMAT_STRING_ROLE).addSourceLine(this));
+        }
+    }
+
+    private boolean contains(String item, String... items) {
+        return Arrays.asList(items).contains(item);
+    }
+
+    /**
+     * Runs the check per each bad combination (Rule objects) on the provided string.
+     *
+     * @param dateFormat - string to be checked
+     * @return {@code true} if given string matches any bad combination.
+     */
+    private boolean runDateFormatRuleVerify(String dateFormat) {
+        for (Rule rule : new Rule[] {
+            // when "h" or "K" flags found, make sure that it ALSO CONTAINS "a" or "B"
+            new Rule(Arrays.asList("a", "B"), null, true, Arrays.asList("h", "K")),
+
+            // when "H" or "k" flags found, make sure that it DOES NOT CONTAIN "a" or "B"
+            new Rule(Arrays.asList("a", "B"), null, false, Arrays.asList("H", "k")),
+
+            // when "M" or "d" flags found, make sure that it DOES NOT CONTAIN "Y" (unless "w" is provided)
+            new Rule(Collections.singletonList("Y"), "w", false, Arrays.asList("M", "d")),
+
+            // milli-of-day cannot be used together with Hours, Minutes, Seconds
+            new Rule(Arrays.asList("H", "h", "K", "k", "m", "s"), null, false, Arrays.asList("A", "N")),
+
+            // milli-of-day and nano-of-day cannot be used together
+            new Rule(Collections.singletonList("A"), null, false, Collections.singletonList("N")),
+            new Rule(Collections.singletonList("N"), null, false, Collections.singletonList("A")),
+
+            // fraction-of-second and nano-of-second cannot be used together
+            new Rule(Collections.singletonList("S"), null, false, Collections.singletonList("n")),
+            new Rule(Collections.singletonList("n"), null, false, Collections.singletonList("S")),
+
+            // am-pm marker and period-of-day cannot be used together
+            new Rule(Collections.singletonList("a"), null, false, Collections.singletonList("B")),
+            new Rule(Collections.singletonList("B"), null, false, Collections.singletonList("a")),
+
+            // year and year-of-era cannot be used together
+            new Rule(Collections.singletonList("u"), null, false, Collections.singletonList("y")),
+            new Rule(Collections.singletonList("y"), null, false, Collections.singletonList("u"))
+        }) {
+            if (rule.verify(dateFormat)) {
+                return true;
+            }
+        }
+        return false;
```

Comment:
The usage of streams could result in a more elegant code here as well.

Reply:
fixed